scripts/presubmit-tests.sh

```diff
@@ -20,111 +20,206 @@
+function list_changed_files() {
+  /workspace/githubhelper -list-changed-files
+}
+
+# Initialize flags and context for presubmit tests:
+# CHANGED_FILES, IS_PRESUBMIT_EXEMPT_PR and IS_DOCUMENTATION_PR.
+function initialize_environment() {
+  CHANGED_FILES=""
+  IS_PRESUBMIT_EXEMPT_PR=0
+  IS_DOCUMENTATION_PR=0
+  (( ! IS_PRESUBMIT )) && return
+  CHANGED_FILES="$(list_changed_files)"
+  if [[ -n "${CHANGED_FILES}" ]]; then
+    echo -e "Changed files in commit ${PULL_PULL_SHA}:\n${CHANGED_FILES}"
+    local no_presubmit_files="${NO_PRESUBMIT_FILES[*]}"
+    pr_only_contains "${no_presubmit_files}" && IS_PRESUBMIT_EXEMPT_PR=1
+    pr_only_contains "\.md ${no_presubmit_files}" && IS_DOCUMENTATION_PR=1
+  else
+    header "NO CHANGED FILES REPORTED, ASSUMING IT'S AN ERROR AND RUNNING TESTS ANYWAY"
   fi
+  readonly CHANGED_FILES
+  readonly IS_DOCUMENTATION_PR
+  readonly IS_PRESUBMIT_EXEMPT_PR
 }
 
-# Run build tests.
-# If there's no `build_tests` function, do:
-# * `go build` on the entire repo
-# * run `/hack/verify-codegen.sh`
-# * check licenses in `/cmd`
+# Display a pass/fail banner for a test group.
+# Parameters: $1 - test group name (e.g., build)
+#             $2 - result (0=passed, 1=failed)
+function results_banner() {
+  local result
+  [[ $2 -eq 0 ]] && result="PASSED" || result="FAILED"
+  header "$1 tests ${result}"
+}
+
+# Run build tests. If there's no `build_tests` function, run the default
+# build test runner.
 function run_build_tests() {
   (( ! RUN_BUILD_TESTS )) && return 0
   header "Running build tests"
-  if function_exists build_tests; then
-    build_tests
-    return
+  local failed=0
+  # Run pre-build tests, if any
+  if function_exists pre_build_tests; then
+    pre_build_tests || failed=1
+  fi
+  # Don't run build tests if pre-build tests failed
+  if (( ! failed )); then
+    if function_exists build_tests; then
+      build_tests || failed=1
+    else
+      default_build_test_runner || failed=1
+    fi
   fi
+  # Don't run post-build tests if pre/build tests failed
+  if function_exists post_build_tests; then
+    post_build_tests || failed=1
+  fi
+  results_banner "Build" ${failed}
+  return ${failed}
+}
+
+# Default build test runner that:
+# * lint and link check markdown files
+# * `go build` on the entire repo
+# * run `/hack/verify-codegen.sh` (if it exists)
+# * check licenses in `/cmd` (if it exists)
+function default_build_test_runner() {
   local failed=0
+  local mdfiles="$(echo "${CHANGED_FILES}" | grep \.md$)"
+  if [[ -n "${mdfiles}" ]]; then
+    subheader "Linting the markdown files"
+    lint_markdown ${mdfiles} || failed=1
+    subheader "Checking links in the markdown files"
+    check_links_in_markdown ${mdfiles} || failed=1
+  fi
+  # For documentation PRs, just check the md files
+  (( IS_DOCUMENTATION_PR )) && return ${failed}
+  # Ensure all the code builds
+  subheader "Checking that go code builds"
   go build -v ./... || failed=1
```

Comment:
I understand this is not part of this PR, just wondering while testing locally, would `go build -v ./...` scan everything under `${GOPATH}/src/github.com` as current `pwd` is `${REPO_ROOT_DIR}`

Reply:
No, it will scan everything under `${GOPATH}/src/github.com/YOUR_REPO_NAME_HERE`, which is `${REPO_ROOT_DIR}`.